contracts/libraries/SwapCalculator.sol

```diff
@@ -0,0 +1,379 @@
+
+        // Calculate the swap
+        uint256 y = balances[outputIndex].sub(outputAmount);
+        uint256 x = getX(a, inputIndex, outputIndex, y, balances);
+        inputAmount = x.sub(balances[inputIndex]).add(1);
+    }
+
+    /**
+     * @notice Calculate the relative price between two assets in given setup of balances and A
+     * @dev Uses 1e18 precision for balances, 1e2 for A
+     * @param balances array of balances
+     * @param a A parameter to be used in the calculation
+     * @param inputIndex index of the input token in the pool
+     * @param outputIndex index of the output token in the pool
+     * @return price relative price of output tokens per one input token
+     */
+    function relativePriceCustom(
+        uint256[] memory balances,
+        uint256 a,
+        uint256 inputIndex,
+        uint256 outputIndex
+    ) public pure returns (uint256 price) {
+        return
+            calculateSwapOutputCustom(
+                balances,
+                a,
+                0,
+                inputIndex,
+                outputIndex,
+                BALANCE_PRECISION
+            );
+    }
+
+    /**
+     * @notice Add and registers a new pool. This function exist to cache decimal information.
+     * @param pool address of a pool contract that implements ISwap
+     */
+    function addPool(address pool) external payable {
+        uint256[] memory decimalsArr = new uint256[](MAX_TOKENS_LENGTH);
+
+        for (uint256 i = 0; i < MAX_TOKENS_LENGTH; i++) {
+            try ISwap(pool).getToken(uint8(i)) returns (IERC20 token) {
+                require(address(token) != address(0), "Token invalid");
+                decimalsArr[i] = IERC20Decimals(address(token)).decimals();
+            } catch {
+                assembly {
+                    mstore(decimalsArr, sub(mload(decimalsArr), sub(8, i)))
+                }
+                break;
+            }
+        }
+
+        require(decimalsArr.length > 0, "Must call addPool() first");
+        storedDecimals[pool] = decimalsArr;
+    }
+
+    /**
+     * @notice Get D, the StableSwap invariant, based on a set of balances and a particular A.
+     * @param xp a precision-adjusted set of pool balances. Array should be the same cardinality
+     * as the pool.
+     * @param a the amplification coefficient * n * (n - 1) in A_PRECISION.
+     * See the StableSwap paper for details
+     * @return the invariant, at the precision of the pool
+     */
+    function getD(uint256[] memory xp, uint256 a)
+        internal
+        pure
+        returns (uint256)
+    {
+        uint256 numTokens = xp.length;
+        uint256 s;
+        for (uint256 i = 0; i < numTokens; i++) {
+            s = s.add(xp[i]);
+        }
+        if (s == 0) {
+            return 0;
+        }
+
+        uint256 prevD;
+        uint256 d = s;
+        uint256 nA = a.mul(numTokens);
+
+        for (uint256 i = 0; i < MAX_LOOP_LIMIT; i++) {
+            uint256 dP = d;
```

Comment:
define variables outside of loops for gas savings! I don't know why!

Reply:
Hmm maybe there is an additional condition. Moving `dP` definition outside the loop actually increases the gas cost... 😕